File: agar-clone/server-python/protocol.py
"""protocol.py -- binary (de)serialization matching SPEC §12 EXACTLY.

Byte-for-byte mirror of server/protocol.js:
  - Every multibyte number is LITTLE-ENDIAN.
  - 1-byte packet id at offset 0 of every frame (both directions).
  - Positions are absolute u16, sizes (radii) u16, leaderboard masses u32.
  - Strings = [u8 length][length x u16 code unit] UTF-16 (charCodeAt), clamped
    to NICK_MAX code units.
  - Quantization is WIRE-ONLY; the sim runs on full-precision floats.
"""

import struct
import constants as C
from physics import js_round

# --- packet ids ------------------------------------------------------------
S2C_WELCOME = 0
S2C_SNAPSHOT = 1
S2C_LEADERBOARD = 2
S2C_PONG = 3
S2C_DEATH = 4
S2C_VERSION_OUTDATED = 255

C2S_SET_NICK = 0
C2S_PING = 1
C2S_INPUT_TARGET = 2
C2S_SPLIT = 3
C2S_EJECT = 4
C2S_RESPAWN = 5
C2S_HANDSHAKE = 255

# cell-block flag bits (SPEC §12.2)
FLAG_MINE = 1 << 0
FLAG_NAME = 1 << 1
FLAG_SPLIT = 1 << 2


def _clamp(v, lo, hi):
    return lo if v < lo else (hi if v > hi else v)


def enc_pos(v):
    return _clamp(js_round(v), 0, 65535)


def enc_size(v):
    return _clamp(js_round(v), 0, 65535)


# --- little-endian writer ---------------------------------------------------
class Writer:
    def __init__(self, size=64):
        self.buf = bytearray()

    def u8(self, v):
        self.buf.append(v & 0xFF)
        return self

    def u16(self, v):
        self.buf += struct.pack("<H", v & 0xFFFF)
        return self

    def u32(self, v):
        self.buf += struct.pack("<I", v & 0xFFFFFFFF)
        return self

    def str(self, s):
        # [u8 length][length x u16 code unit] UTF-16-LE; clamp to NICK_MAX units.
        s = s or ""
        units = s.encode("utf-16-le")  # 2 bytes per UTF-16 code unit
        n = len(units) // 2
        if n > C.NICK_MAX:
            n = C.NICK_MAX
            units = units[: n * 2]
        self.buf.append(n & 0xFF)
        self.buf += units
        return self

    def done(self):
        return bytes(self.buf)


# --- reader for client->server frames --------------------------------------
class Reader:
    def __init__(self, buf):
        self.buf = buf
        self.pos = 0

    def u8(self):
        v = self.buf[self.pos]
        self.pos += 1
        return v

    def u16(self):
        v = struct.unpack_from("<H", self.buf, self.pos)[0]
        self.pos += 2
        return v

    def u32(self):
        v = struct.unpack_from("<I", self.buf, self.pos)[0]
        self.pos += 4
        return v

    def str(self):
        length = self.u8()
        raw = self.buf[self.pos : self.pos + 2 * length]
        self.pos += 2 * length
        return raw.decode("utf-16-le", errors="replace")

    def remaining(self):
        return len(self.buf) - self.pos


# ===========================================================================
# Server -> Client encoders
# ===========================================================================

def welcome(your_id, server_tick):
    w = Writer(16)
    (w.u8(S2C_WELCOME)
      .u32(your_id)
      .u16(C.WORLD_SIZE)
      .u16(C.WORLD_SIZE)
      .u32(server_tick))
    return w.done()


def snapshot(server_tick, eats, cells, foods, viruses, ejects, removes):
    w = Writer(2048)
    w.u8(S2C_SNAPSHOT).u32(server_tick)

    w.u16(len(eats))
    for e in eats:
        w.u32(e["eaterId"]).u32(e["eatenId"])

    w.u16(len(cells))
    for c in cells:
        (w.u32(c["id"])
          .u32(c["ownerId"])
          .u16(enc_pos(c["x"]))
          .u16(enc_pos(c["y"]))
          .u16(enc_size(c["size"]))
          .u8(c["hue"] & 0xFF)
          .u8(c["flags"] & 0xFF))
        if c["flags"] & FLAG_NAME:
            w.str(c["name"] or "")

    w.u16(len(foods))
    for f in foods:
        w.u32(f["id"]).u16(enc_pos(f["x"])).u16(enc_pos(f["y"])).u8(f["hue"] & 0xFF)

    w.u16(len(viruses))
    for v in viruses:
        w.u32(v["id"]).u16(enc_pos(v["x"])).u16(enc_pos(v["y"])).u16(enc_size(v["size"]))

    w.u16(len(ejects))
    for e in ejects:
        w.u32(e["id"]).u16(enc_pos(e["x"])).u16(enc_pos(e["y"])).u8(e["hue"] & 0xFF)

    w.u16(len(removes))
    for rid in removes:
        w.u32(rid)

    return w.done()


def leaderboard(rows, your_rank):
    w = Writer(256)
    w.u8(S2C_LEADERBOARD).u8(len(rows))
    for r in rows:
        w.u32(r["id"]).u32(r["mass"] & 0xFFFFFFFF).str(r["name"] or "")
    w.u8(_clamp(int(your_rank), 0, 255))
    return w.done()


def pong(client_ms, server_tick):
    w = Writer(9)
    w.u8(S2C_PONG).u32(client_ms & 0xFFFFFFFF).u32(server_tick & 0xFFFFFFFF)
    return w.done()


def death(final_mass):
    w = Writer(5)
    w.u8(S2C_DEATH).u32(max(0, int(final_mass)))
    return w.done()


def version_outdated():
    return bytes([S2C_VERSION_OUTDATED])


# ===========================================================================
# Client -> Server decoder
# ===========================================================================

def decode_client(buf):
    if not buf or len(buf) < 1:
        return None
    r = Reader(buf)
    pid = r.u8()
    if pid == C2S_HANDSHAKE:
        version = r.u32() if len(buf) >= 5 else 0
        return {"type": "handshake", "version": version}
    if pid == C2S_SET_NICK:
        if len(buf) < 2:
            return {"type": "nick", "name": ""}
        return {"type": "nick", "name": r.str()}
    if pid == C2S_PING:
        client_ms = r.u32() if len(buf) >= 5 else 0
        return {"type": "ping", "clientMs": client_ms}
    if pid == C2S_INPUT_TARGET:
        if len(buf) < 5:
            return None
        x = r.u16()
        y = r.u16()
        return {"type": "target", "x": x, "y": y}
    if pid == C2S_SPLIT:
        return {"type": "split"}
    if pid == C2S_EJECT:
        return {"type": "eject"}
    if pid == C2S_RESPAWN:
        return {"type": "respawn"}
    return {"type": "unknown", "id": pid}

File: agar-clone/server-go/protocol.go
package main

import "encoding/binary"

// protocol.go — binary (de)serialization matching SPEC §12 and server/protocol.js
// byte-for-byte. Every multibyte number is LITTLE-ENDIAN. 1-byte packet id at
// offset 0 of every frame (both directions). World positions/sizes are u16,
// leaderboard masses u32. Strings = [u8 length][length × u16 code unit] UTF-16
// (charCodeAt semantics); names <= NickMax code units, server clamps.

// --- packet ids (Server -> Client) ---
const (
	S2CWelcome         = 0
	S2CSnapshot        = 1
	S2CLeaderboard     = 2
	S2CPong            = 3
	S2CDeath           = 4
	S2CVersionOutdated = 255
)

// --- packet ids (Client -> Server) ---
const (
	C2SSetNick     = 0
	C2SPing        = 1
	C2SInputTarget = 2
	C2SSplit       = 3
	C2SEject       = 4
	C2SRespawn     = 5
	C2SHandshake   = 255
)

// cell-block flag bits (SPEC §12.2)
const (
	FlagMine  = 1 << 0
	FlagName  = 1 << 1
	FlagSplit = 1 << 2
)

// nick is a sequence of UTF-16 code units, matching the JS string model.
type nick []uint16

func clampI(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

// absolute world coord -> u16
func encPos(v float64) uint16 { return uint16(clampI(jsRound(v), 0, 65535)) }

// radius -> u16
func encSize(v float64) uint16 { return uint16(clampI(jsRound(v), 0, 65535)) }

// --- tiny growable little-endian writer ---
type Writer struct {
	buf []byte
}

func newWriter(size int) *Writer { return &Writer{buf: make([]byte, 0, size)} }

func (w *Writer) u8(v int) *Writer {
	w.buf = append(w.buf, byte(v&0xff))
	return w
}
func (w *Writer) u16(v int) *Writer {
	var b [2]byte
	binary.LittleEndian.PutUint16(b[:], uint16(v&0xffff))
	w.buf = append(w.buf, b[:]...)
	return w
}
func (w *Writer) u16u(v uint16) *Writer {
	var b [2]byte
	binary.LittleEndian.PutUint16(b[:], v)
	w.buf = append(w.buf, b[:]...)
	return w
}
func (w *Writer) u32(v uint32) *Writer {
	var b [4]byte
	binary.LittleEndian.PutUint32(b[:], v)
	w.buf = append(w.buf, b[:]...)
	return w
}

// str: [u8 length][length × u16 code unit] UTF-16; clamps to NickMax code units.
func (w *Writer) str(s nick) *Writer {
	if len(s) > NickMax {
		s = s[:NickMax]
	}
	w.u8(len(s) & 0xff)
	for _, cu := range s {
		w.u16u(cu)
	}
	return w
}

func (w *Writer) done() []byte { return w.buf }

// --- reader for client->server frames ---
type Reader struct {
	buf []byte
	pos int
}

func newReader(b []byte) *Reader { return &Reader{buf: b} }

func (r *Reader) u8() int {
	v := int(r.buf[r.pos])
	r.pos++
	return v
}
func (r *Reader) u16() int {
	v := int(binary.LittleEndian.Uint16(r.buf[r.pos:]))
	r.pos += 2
	return v
}
func (r *Reader) u32() uint32 {
	v := binary.LittleEndian.Uint32(r.buf[r.pos:])
	r.pos += 4
	return v
}
func (r *Reader) str() nick {
	n := r.u8()
	s := make(nick, 0, n)
	for i := 0; i < n; i++ {
		s = append(s, uint16(r.u16()))
	}
	return s
}

// ===========================================================================
// Server -> Client encoders
// ===========================================================================

// id 0 WELCOME: [u8 0][u32 yourPlayerId][u16 worldW][u16 worldH][u32 serverTick]
func encWelcome(yourID uint32, serverTick uint32) []byte {
	w := newWriter(16)
	w.u8(S2CWelcome).u32(yourID).u16(WorldSize).u16(WorldSize).u32(serverTick)
	return w.done()
}

// snapshot wire structs (already culled to a viewer's AoI by main.go)
type sCell struct {
	id, ownerID uint32
	x, y, size  float64
	hue         int
	flags       int
	name        nick // only written when FlagName is set
}
type sFood struct {
	id   uint32
	x, y float64
	hue  int
}
type sVirus struct {
	id         uint32
	x, y, size float64
}
type sEject struct {
	id   uint32
	x, y float64
	hue  int
}
type sEat struct {
	eaterID, eatenID uint32
}

// id 1 SNAPSHOT (SPEC §12.2)
func encSnapshot(serverTick uint32, eats []sEat, cells []sCell, foods []sFood, viruses []sVirus, ejects []sEject, removes []uint32) []byte {
	w := newWriter(2048)
	w.u8(S2CSnapshot).u32(serverTick)

	w.u16(len(eats))
	for _, e := range eats {
		w.u32(e.eaterID).u32(e.eatenID)
	}

	w.u16(len(cells))
	for _, c := range cells {
		w.u32(c.id).u32(c.ownerID).
			u16u(encPos(c.x)).u16u(encPos(c.y)).u16u(encSize(c.size)).
			u8(c.hue & 0xff).u8(c.flags & 0xff)
		if c.flags&FlagName != 0 {
			w.str(c.name)
		}
	}

	w.u16(len(foods))
	for _, f := range foods {
		w.u32(f.id).u16u(encPos(f.x)).u16u(encPos(f.y)).u8(f.hue & 0xff)
	}

	w.u16(len(viruses))
	for _, v := range viruses {
		w.u32(v.id).u16u(encPos(v.x)).u16u(encPos(v.y)).u16u(encSize(v.size))
	}

	w.u16(len(ejects))
	for _, e := range ejects {
		w.u32(e.id).u16u(encPos(e.x)).u16u(encPos(e.y)).u8(e.hue & 0xff)
	}

	w.u16(len(removes))
	for _, id := range removes {
		w.u32(id)
	}

	return w.done()
}

// id 2 LEADERBOARD: [u8 2][u8 n]{ u32 playerId, u32 mass, str name }[u8 yourRank]
type lbRow struct {
	id   uint32
	mass uint32
	name nick
}

func encLeaderboard(rows []lbRow, yourRank int) []byte {
	w := newWriter(256)
	w.u8(S2CLeaderboard).u8(len(rows))
	for _, r := range rows {
		w.u32(r.id).u32(r.mass).str(r.name)
	}
	w.u8(clampI(yourRank, 0, 255))
	return w.done()
}

// id 3 PONG: [u8 3][u32 clientMs][u32 serverTick]
func encPong(clientMs uint32, serverTick uint32) []byte {
	w := newWriter(9)
	w.u8(S2CPong).u32(clientMs).u32(serverTick)
	return w.done()
}

// id 4 DEATH: [u8 4][u32 finalMass]
func encDeath(finalMass int) []byte {
	if finalMass < 0 {
		finalMass = 0
	}
	w := newWriter(5)
	w.u8(S2CDeath).u32(uint32(finalMass))
	return w.done()
}

// id 255 VERSION_OUTDATED: no payload
func encVersionOutdated() []byte { return []byte{S2CVersionOutdated} }

// ===========================================================================
// Client -> Server decoder
// ===========================================================================

type clientMsg struct {
	typ     string
	version uint32
	name    nick
	clientMs uint32
	x, y    int
	id      int
}

func decodeClient(buf []byte) *clientMsg {
	if len(buf) < 1 {
		return nil
	}
	r := newReader(buf)
	id := r.u8()
	switch id {
	case C2SHandshake:
		var v uint32
		if len(buf) >= 5 {
			v = r.u32()
		}
		return &clientMsg{typ: "handshake", version: v}
	case C2SSetNick:
		if len(buf) < 2 {
			return &clientMsg{typ: "nick", name: nick{}}
		}
		return &clientMsg{typ: "nick", name: r.str()}
	case C2SPing:
		var ms uint32
		if len(buf) >= 5 {
			ms = r.u32()
		}
		return &clientMsg{typ: "ping", clientMs: ms}
	case C2SInputTarget:
		if len(buf) < 5 {
			return nil
		}
		x := r.u16()
		y := r.u16()
		return &clientMsg{typ: "target", x: x, y: y}
	case C2SSplit:
		return &clientMsg{typ: "split"}
	case C2SEject:
		return &clientMsg{typ: "eject"}
	case C2SRespawn:
		return &clientMsg{typ: "respawn"}
	default:
		return &clientMsg{typ: "unknown", id: id}
	}
}
